src/crush/CrushWrapper.cc

```diff
@@ -3872,18 +3872,20 @@ int CrushWrapper::_choose_type_stack(
   //      are careful to pick one that has the same parent.)
   vector<set<int>> underfull_buckets; // level -> set of buckets with >0 underfull item(s)
   underfull_buckets.resize(stack.size() - 1);
-  for (auto osd : underfull) {
-    int item = osd;
-    for (int j = (int)stack.size() - 2; j >= 0; --j) {
-      int type = stack[j].first;
-      item = get_parent_of_type(item, type, rule);
-      ldout(cct, 10) << __func__ << " underfull " << osd << " type " << type
-		     << " is " << item << dendl;
-      if (!subtree_contains(root_bucket, item)) {
-        ldout(cct, 20) << __func__ << " not in root subtree " << root_bucket << dendl;
-        continue;
+  for (auto& ls : { underfull, more_underfull }) {
+    for (auto osd : ls) {
+      int item = osd;
+      for (int j = (int)stack.size() - 2; j >= 0; --j) {
+        int type = stack[j].first;
+        item = get_parent_of_type(item, type, rule);
+        ldout(cct, 10) << __func__ << " underfull " << osd << " type " << type
+		       << " is " << item << dendl;
+        if (!subtree_contains(root_bucket, item)) {
+          ldout(cct, 20) << __func__ << " not in root subtree " << root_bucket << dendl;
+          continue;
+        }
+        underfull_buckets[j].insert(item);
```

Comment:
@dzafman You need to add more_underfull into underfull_buckets here, otherwise the swap between overfull and more_underfull would not really happen. 

See https://github.com/ceph/ceph/blob/master/src/crush/CrushWrapper.cc#L3992  and below.

Reply:
@xiexingguo I still don't understand.  With the current code here is log output showing a case where more_underfull came into play in the old code.

```
020-10-27T18:35:11.561+0000 7ff31415dc40 10 _choose_type_stack more underfull pos 0 was 1904 considering 1887
2020-10-27T18:35:11.561+0000 7ff31415dc40 20 _choose_type_stack   not in subtree -594
2020-10-27T18:35:11.561+0000 7ff31415dc40 10 _choose_type_stack more underfull pos 0 was 1904 considering 1888
2020-10-27T18:35:11.561+0000 7ff31415dc40 20 _choose_type_stack   not in subtree -594
2020-10-27T18:35:11.561+0000 7ff31415dc40 10 _choose_type_stack more underfull pos 0 was 1904 considering 1895
2020-10-27T18:35:11.561+0000 7ff31415dc40 10 _choose_type_stack pos 0 replace 1904 -> 1895
...
2020-10-27T18:35:11.561+0000 7ff31415dc40 10 _choose_type_stack pos 0 was 166 considering 406
2020-10-27T18:35:11.561+0000 7ff31415dc40 10 _choose_type_stack pos 0 replace 166 -> 406
2020-10-27T18:35:11.561+0000 7ff31415dc40 10 _choose_type_stack end of orig, break 1
2020-10-27T18:35:11.561+0000 7ff31415dc40 10 _choose_type_stack end of orig, break 2
2020-10-27T18:35:11.561+0000 7ff31415dc40 10 _choose_type_stack  w <- [707,673,394,2524,2265,2475,1895,2193,2089,344,406] was [-53,-51,-47,-456,-423,-450,-594,-633,-621,-41,-11,-3]
2020-10-27T18:35:11.561+0000 7ff31415dc40 10 try_remap_rule step 3 w [707,673,394,2524,2265,2475,1895,2193,2089,344,406]
2020-10-27T18:35:11.561+0000 7ff31415dc40 10  emit [707,673,394,2524,2265,2475,1895,2193,2089,344,406]
```

Here are the pg-upmap-items generated that include 1895.  The last one corresponds with this debug_crush=20 log above.

```
ceph osd pg-upmap-items 21.137e 1497 1510 1889 1895
ceph osd pg-upmap-items 21.39e 1075 1079 1889 1895
ceph osd pg-upmap-items 7.11f2 1895 1894 1247 1231 137 477
ceph osd pg-upmap-items 21.b36 2309 2312 1461 1466 88 386 1889 1895
ceph osd pg-upmap-items 21.1bd6 1904 1895
```